lib/Sema/BuilderTransform.cpp

```diff
@@ -1619,16 +1621,24 @@ ConstraintSystem::matchFunctionBuilder(
 
   // Pre-check the body: pre-check any expressions in it and look
   // for return statements.
-  auto request = PreCheckFunctionBuilderRequest{fn};
+  auto request =
+      PreCheckFunctionBuilderRequest{{fn, /*SuppressDiagnostics=*/true}};
   switch (evaluateOrDefault(getASTContext().evaluator, request,
                             FunctionBuilderBodyPreCheck::Error)) {
   case FunctionBuilderBodyPreCheck::Okay:
     // If the pre-check was okay, apply the function-builder transform.
     break;
 
-  case FunctionBuilderBodyPreCheck::Error:
-    // If the pre-check had an error, flag that.
-    return getTypeMatchFailure(locator);
+  case FunctionBuilderBodyPreCheck::Error: {
+    if (!shouldAttemptFixes())
+      return getTypeMatchFailure(locator);
+
+    if (recordFix(IgnoreInvalidFunctionBuilderBody::create(
```

Comment:
Instead of suppressing diagnostics and doing another walk later to diagnose the pre-check errors in `IgnoreInvalidFunctionBuilderBody`, can we emit errors here and find another way to tell the constraint system that it has already diagnosed an error? It seems strange to me that we're calling `preCheckExpression` from diagnostics code just to delay emitting the errors until `applySolutionWithFixes`.

`DefineMemberBasedOnUse` does a similar thing for errors that were diagnosed at a declaration elsewhere. The fix is still recorded to tell the constraint system that there was an error, but it doesn't diagnose anything itself. We could also potentially still record the function builder transformation and note that it had an error.

Reply:
Oh, I realized this is because it's possible for the solver not pick the solution that has the function builder transformation applied in the case of a closure. Nevermind!